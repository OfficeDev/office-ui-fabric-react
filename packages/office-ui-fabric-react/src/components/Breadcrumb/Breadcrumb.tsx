import * as React from 'react';

import {
  IBreadcrumbProps,
  IBreadcrumbStyles,
  IBreadcrumbStyleProps
} from './Breadcrumb.props';
import { BreadcrumbBase } from './Breadcrumb.base';
import { styled } from '../../Styling';
import { ICrumbStyleProps, ICrumbStyles } from './Crumb.props';
import { getStyles } from './Breadcrumb.styles';
import { getStyles as getCrumbStyles } from './Crumb.styles';

// Create a Breadcrumb variant which uses these default styles.
export const Breadcrumb = styled(
  BreadcrumbBase,
  {
    getStyles,
    getCrumbStyles
  }
);
