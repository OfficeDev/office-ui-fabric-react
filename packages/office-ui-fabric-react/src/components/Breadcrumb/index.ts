export * from './Breadcrumb.base';
export * from './Breadcrumb.props';
export * from './Breadcrumb';
