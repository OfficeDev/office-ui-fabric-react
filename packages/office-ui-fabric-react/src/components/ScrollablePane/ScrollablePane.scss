@import '../../common/common';

//
// Office UI Fabric
// --------------------------------------------------
// Panel styles


.root {
}

.scrollCopy {
  position: absolute;
  pointer-events: none;
}

.scrollRegion {
}

.contentArea {
}

.stickyContainer {
  position: relative;
  max-height: inherit;
  height: 100%;
}

.layerHostTop {
  background: $ms-color-white;
  position: absolute;
  // visibility: hidden;
  top: 0;
  width: 100%;
  pointer-events: auto;
}

.layerHostBottom {
  background: $ms-color-white;
  position: absolute;
  // visibility: hidden;
  bottom: 0;
  width: 100%;
  pointer-events: auto;
}