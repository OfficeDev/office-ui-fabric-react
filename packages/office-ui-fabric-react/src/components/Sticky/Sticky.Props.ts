import * as React from 'react';
import { Sticky } from './Sticky';

export interface IStickyProps extends React.Props<Sticky> {
  /**
   * Optional callback to access the IDialog interface. Use this instead of ref for accessing
   * the public methods and properties of the component.
   */
  componentRef?: (component: IStickyProps) => void;

  /**
   * Class name to apply to the sticky element if component is sticky.
  */
  stickyClassName?: string;
}