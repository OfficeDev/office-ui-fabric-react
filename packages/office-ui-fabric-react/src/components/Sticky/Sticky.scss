@import '../../common/common';

.isSticky {
  position: absolute;
  -webkit-overflow-scrolling: touch;
  -webkit-transform: translateZ(0);
  transform: translateZ(0);
  -webkit-backface-visibility: hidden;
  backface-visibility: hidden;
  z-index: 1;
}