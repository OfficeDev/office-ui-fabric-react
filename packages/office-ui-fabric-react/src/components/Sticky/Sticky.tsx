import * as PropTypes from 'prop-types';
import * as React from 'react';
import { BaseComponent } from '../../Utilities';
import {
  IScrollablePaneContext,
  ScrollablePaneContext,
  PlaceholderPosition,
  StickyContainerBehaviorType
} from '../ScrollablePane/ScrollablePane.types';
import { IStickyProps, StickyPositionType } from './Sticky.types';

export interface IStickyState {
  isStickyTop: boolean;
  isStickyBottom: boolean;
  distanceFromTop?: number;
}

export interface IStickyContext {
  scrollablePane: PropTypes.Requireable<object>;
}

export class Sticky extends BaseComponent<IStickyProps, IStickyState> {
  public static defaultProps: IStickyProps = {
    stickyPosition: StickyPositionType.Both,
    isScrollSynced: true
  };

  public static contextType = ScrollablePaneContext;

  private _root = React.createRef<HTMLDivElement>();
  private _stickyContentTop = React.createRef<HTMLDivElement>();
  private _stickyContentBottom = React.createRef<HTMLDivElement>();
  private _nonStickyContent = React.createRef<HTMLDivElement>();
  private _placeHolder = React.createRef<HTMLDivElement>();
  private _activeElement: HTMLElement | undefined;

  constructor(props: IStickyProps) {
    super(props);
    this.state = {
      isStickyTop: false,
      isStickyBottom: false,
      distanceFromTop: undefined
    };
    this._activeElement = undefined;
  }

  public get root(): HTMLDivElement | null {
    return this._root.current;
  }

  public get placeholder(): HTMLDivElement | null {
    return this._placeHolder.current;
  }

  public get stickyContentTop(): HTMLDivElement | null {
    return this._stickyContentTop.current;
  }

  public get stickyContentBottom(): HTMLDivElement | null {
    return this._stickyContentBottom.current;
  }

  public get nonStickyContent(): HTMLDivElement | null {
    return this._nonStickyContent.current;
  }

  public get canStickyTop(): boolean {
    return this.props.stickyPosition === StickyPositionType.Both || this.props.stickyPosition === StickyPositionType.Header;
  }

  public get canStickyBottom(): boolean {
    return this.props.stickyPosition === StickyPositionType.Both || this.props.stickyPosition === StickyPositionType.Footer;
  }

  public syncScroll = (container: HTMLElement): void => {
    const { nonStickyContent } = this;
    const { isStickyBottom, isStickyTop } = this.state;
    const { scrollablePane } = this.context;
    // scroll sync is needed only if current state is sticky
    if (!scrollablePane || !this.props.isScrollSynced || !(isStickyBottom || isStickyTop)) {
      return;
    }
    const containerScrollLeft = scrollablePane.getScrollPosition(true /** horizontal */);
    if (isStickyTop && !scrollablePane.usePlaceholderForSticky('top' /** StickyContentTop */) && this.stickyContentTop) {
      this.stickyContentTop.children[0].scrollLeft = containerScrollLeft;
    } else if (isStickyBottom && !scrollablePane.usePlaceholderForSticky('bottom' /** StickyContentBottom */) && this.stickyContentBottom) {
      this.stickyContentBottom.children[0].scrollLeft = containerScrollLeft;
    } else if (nonStickyContent) {
      nonStickyContent.scrollLeft = containerScrollLeft;
    }
  };

  public componentDidMount(): void {
    const { scrollablePane } = this._getContext();

    if (!scrollablePane) {
      return;
    }

    scrollablePane.subscribe(this._onScrollEvent);
    scrollablePane.addSticky(this);
  }

  public componentWillUnmount(): void {
    const { scrollablePane } = this._getContext();

    if (!scrollablePane) {
      return;
    }

    scrollablePane.unsubscribe(this._onScrollEvent);
    scrollablePane.removeSticky(this);
  }

  public componentDidUpdate(prevProps: IStickyProps, prevState: IStickyState): void {
    const { scrollablePane } = this._getContext();

    if (!scrollablePane) {
      return;
    }

    const { isStickyBottom, isStickyTop, distanceFromTop } = this.state;
    if (prevState.distanceFromTop !== distanceFromTop) {
      scrollablePane.sortSticky(this, true /*sortAgain*/);
    }
    if (prevState.isStickyTop !== isStickyTop || prevState.isStickyBottom !== isStickyBottom) {
      if (this._activeElement) {
        this._activeElement.focus();
      }
      scrollablePane.updateStickyRefHeights();
    }
    // horizontal scroll has to be synced only if component is sticky
    if ((isStickyBottom || isStickyTop) && scrollablePane.getScrollPosition(true /** horizontal */)) {
      // Sync Sticky scroll position with content container on each update
      scrollablePane.syncScrollSticky(this);
    }
  }

  public shouldComponentUpdate(nextProps: IStickyProps, nextState: IStickyState): boolean {
    if (!this.context.scrollablePane) {
      return true;
    }

    const { isStickyTop, isStickyBottom, distanceFromTop } = this.state;

    return (isStickyTop !== nextState.isStickyTop ||
      isStickyBottom !== nextState.isStickyBottom ||
      this.props.stickyPosition !== nextProps.stickyPosition ||
      this.props.children !== nextProps.children ||
      distanceFromTop !== nextState.distanceFromTop ||
      this._isOffsetHeightDifferent()) as boolean;
  }

  public render(): JSX.Element {
    const { isStickyTop, isStickyBottom } = this.state;
    const { stickyClassName, children } = this.props;

    if (!this.context.scrollablePane) {
      return <div>{this.props.children}</div>;
    }

    return (
      <div ref={this._root}>
        {this.canStickyTop && this._getStickyContent('top', isStickyTop)}
        {this.canStickyBottom && this._getStickyContent('bottom', isStickyBottom)}
        <div style={this._getNonStickyPlaceholderHeightAndWidth()} ref={this._placeHolder}>
          <div
            ref={this._nonStickyContent}
            className={isStickyTop || isStickyBottom ? stickyClassName : undefined}
            style={this._getContentStyles(isStickyTop || isStickyBottom)}
          >
            {children}
          </div>
        </div>
      </div>
    );
  }

  public addSticky(stickyContent: HTMLDivElement, placeholderPosition: PlaceholderPosition): void {
    const placeholderUsedForStickyContent = this.context.scrollablePane.usePlaceholderForSticky(placeholderPosition);
    if (placeholderUsedForStickyContent && this.nonStickyContent) {
      stickyContent.appendChild(this.nonStickyContent);
    }
  }

  public resetSticky(): void {
    const placeholderUsedForContent = this.context.scrollablePane.usePlaceholderForSticky(this.canStickyTop ? 'top' : 'bottom');
    if (placeholderUsedForContent && this.nonStickyContent && this.placeholder) {
      this.placeholder.appendChild(this.nonStickyContent);
    }
  }

  public setDistanceFromTop(container: HTMLDivElement): void {
    const distanceFromTop = this._getNonStickyDistanceFromTop(container);
    this.setState({ distanceFromTop: distanceFromTop });
  }

  private _getContext = (): IScrollablePaneContext => this.context;

  private _getContentStyles(isSticky: boolean): React.CSSProperties {
    const { scrollablePane } = this.context;
    if (!scrollablePane) {
      return {};
    }
    const isVisible = isSticky
      ? (this.canStickyTop && scrollablePane.usePlaceholderForSticky('top')) ||
        (this.canStickyBottom && scrollablePane.usePlaceholderForSticky('bottom'))
      : true;
    return {
      backgroundColor: this.props.stickyBackgroundColor || this._getBackground(),
      overflow: isSticky ? 'hidden' : '',
      visibility: isVisible ? 'visible' : 'hidden'
    };
  }

  private _getStickyContent(placeholderPosition: PlaceholderPosition, isSticky: boolean): JSX.Element {
    const { stickyClassName, children } = this.props;
    const { scrollablePane } = this.context;
    // decide if actual element is to be replicated or placeholder is be to used.
    const usePlaceholderForStickyContent = scrollablePane.usePlaceholderForSticky(placeholderPosition);
    return (
      <div
        ref={placeholderPosition === 'top' ? this._stickyContentTop : this._stickyContentBottom}
        aria-hidden={!isSticky}
        style={{ pointerEvents: isSticky ? 'auto' : 'none' }}
      >
        {usePlaceholderForStickyContent ? (
          <div style={this._getStickyPlaceholderHeight(isSticky)} />
        ) : (
          <div className={isSticky ? stickyClassName : undefined} style={this._getStickyContentStyles(isSticky)}>
            {children}
          </div>
        )}
      </div>
    );
  }

  private _getStickyContentStyles(isSticky: boolean): React.CSSProperties {
    return {
      visibility: isSticky ? 'visible' : 'hidden',
      pointerEvents: isSticky ? 'auto' : 'none',
      overflow: 'hidden',
      backgroundColor: this.props.stickyBackgroundColor || this._getBackground()
    };
  }

  private _getStickyPlaceholderHeight(isSticky: boolean): React.CSSProperties {
    const height = this.nonStickyContent ? this.nonStickyContent.offsetHeight : 0;
    return {
      visibility: isSticky ? 'hidden' : 'visible',
      height: isSticky ? 0 : height
    };
  }

  private _isOffsetHeightDifferent(): boolean {
    const { scrollablePane } = this.context;
    if (!scrollablePane) {
      return false;
    }

    const usePlaceholderForStickyContentTop = this.canStickyTop && scrollablePane.usePlaceholderForSticky('top');
    const usePlaceholderForStickyContentBottom = this.canStickyBottom && scrollablePane.usePlaceholderForSticky('bottom');

    return (
      (usePlaceholderForStickyContentTop && _isOffsetHeightDifferent(this._nonStickyContent, this._stickyContentTop)) ||
      (usePlaceholderForStickyContentBottom && _isOffsetHeightDifferent(this._nonStickyContent, this._stickyContentBottom)) ||
      ((usePlaceholderForStickyContentBottom || usePlaceholderForStickyContentTop) &&
        _isOffsetHeightDifferent(this._nonStickyContent, this._placeHolder))
    );
  }

  private _getNonStickyPlaceholderHeightAndWidth(): React.CSSProperties {
    const { isStickyTop, isStickyBottom } = this.state;
    const usePlaceholderForStickyTop = this.canStickyTop && this.context.scrollablePane.usePlaceholderForSticky('top');
    const usePlaceholderForStickyBottom = this.canStickyBottom && this.context.scrollablePane.usePlaceholderForSticky('bottom');
    if ((isStickyTop && usePlaceholderForStickyTop) || (isStickyBottom && usePlaceholderForStickyBottom)) {
      let height = 0,
        width = 0;
      // Why is placeHolder width needed?
      // ScrollablePane content--container is reponsible for providing scrollbars depending on content overflow.
      // If the overflow is caused by content of sticky component when it is in non-sticky state,
      // ScrollablePane content--conatiner will provide horizontal scrollbar.
      // If the component becomes sticky, i.e., when state.isStickyTop || state.isStickyBottom becomes true,
      // it's actual content is no more inside ScrollablePane content--container.
      // ScrollablePane content--conatiner will see no need for horizontal scrollbar. (Assuming no other content is causing overflow)
      // The complete content of sticky component will not be viewable.
      // It is necessary to provide a placeHolder of a certain width (height is already being set) in the content--container,
      // to get a horizontal scrollbar & be able to view the complete content of sticky component.
      if (this.nonStickyContent && this.nonStickyContent.firstElementChild) {
        height = this.nonStickyContent.offsetHeight;
        // What value should be substituted for placeHolder width?
        // Assumption:
        //    1. Content inside <Sticky> should always be wrapped in a single div.
        //        <Sticky><div id={'firstElementChild'}>{intended_content}</div><Sticky/>
        //    2. -ve padding, margin, etc. are not be used.
        //    3. scrollWidth of a parent is greater than or equal to max of scrollWidths of it's children and same holds for children.
        // placeHolder width should be computed in the best possible way to prevent overscroll/underscroll.
        width =
          this.nonStickyContent.firstElementChild.scrollWidth +
          ((this.nonStickyContent.firstElementChild as HTMLElement).offsetWidth - this.nonStickyContent.firstElementChild.clientWidth);
      }
      return {
        height: height,
        width: width
      };
    } else {
      return {};
    }
  }

  private _onScrollEvent = (container: HTMLElement, footerStickyContainer: HTMLElement): void => {
    const { scrollablePane } = this.context;
    if (!this.root || !this.nonStickyContent || !scrollablePane) {
      return;
    }
    if (scrollablePane.verifyStickyContainerBehavior(this.canStickyTop ? 'above' : 'below', StickyContainerBehaviorType.StickyAlways)) {
      // 1. ScrollablePane is mounted and has called notifySubscriber
      // 2. stickyAlways has to re-render if mutation could 've affected it's offsetHeight.
      this.setState({
        isStickyTop: this.canStickyTop,
        isStickyBottom: !this.canStickyTop,
        distanceFromTop: 0 // must so that sorting happens.
      });
    } else if (
      !scrollablePane.getUserInteractionStatus() &&
      scrollablePane.verifyStickyContainerBehavior(this.canStickyTop ? 'above' : 'below', StickyContainerBehaviorType.StickyOnScroll)
    ) {
      // user interaction has not started
      // 1. ScrollablePane is mounted and has called notifySubscriber, sort is required.
      // 2. StickyOnScroll has to re-render if mutation could 've affected it's offsetHeight.
      const { isStickyBottom, isStickyTop } = this.state;
      scrollablePane.sortSticky(this, false);
      this.setState({
        isStickyBottom: isStickyBottom,
        isStickyTop: isStickyTop
      });
    } else {
      const distanceFromTop = this._getNonStickyDistanceFromTop(container);
      let isStickyTop = false;
      let isStickyBottom = false;

      if (this.canStickyTop) {
        const distanceToStickTop = distanceFromTop - this._getStickyDistanceFromTop();
        isStickyTop = distanceToStickTop < container.scrollTop;
      }

      // Can sticky bottom if the scrollablePane - total sticky footer height is smaller than the sticky's distance from the top of the pane
      if (this.canStickyBottom && container.clientHeight - footerStickyContainer.offsetHeight <= distanceFromTop) {
        isStickyBottom =
          distanceFromTop - Math.floor(container.scrollTop) >= this._getStickyDistanceFromTopForFooter(container, footerStickyContainer);
      }

      if (
        document.activeElement &&
        this.nonStickyContent.contains(document.activeElement) &&
        (this.state.isStickyTop !== isStickyTop || this.state.isStickyBottom !== isStickyBottom)
      ) {
        this._activeElement = document.activeElement as HTMLElement;
      } else {
        this._activeElement = undefined;
      }

      this.setState({
        isStickyTop: this.canStickyTop && isStickyTop,
        isStickyBottom: isStickyBottom,
        distanceFromTop: distanceFromTop
      });
    }
  };

  private _getStickyDistanceFromTop = (): number => {
    let distance = 0;
    if (this.stickyContentTop) {
      distance = this.stickyContentTop.offsetTop;
    }

    return distance;
  };

  private _getStickyDistanceFromTopForFooter = (container: HTMLElement, footerStickyVisibleContainer: HTMLElement): number => {
    let distance = 0;
    if (this.stickyContentBottom) {
      distance = container.clientHeight - footerStickyVisibleContainer.offsetHeight + this.stickyContentBottom.offsetTop;
    }

    return distance;
  };

  private _getNonStickyDistanceFromTop = (container: HTMLElement): number => {
    let distance = 0;
    let currElem = this.root;

    if (currElem) {
      while (currElem && currElem.offsetParent !== container) {
        distance += currElem.offsetTop;
        currElem = currElem.offsetParent as HTMLDivElement;
      }

      if (currElem && currElem.offsetParent === container) {
        distance += currElem.offsetTop;
      }
    }
    return distance;
  };

  // Gets background of nearest parent element that has a declared background-color attribute
  private _getBackground(): string | undefined {
    if (!this.root) {
      return undefined;
    }

    let curr: HTMLElement = this.root;

    while (
      window.getComputedStyle(curr).getPropertyValue('background-color') === 'rgba(0, 0, 0, 0)' ||
      window.getComputedStyle(curr).getPropertyValue('background-color') === 'transparent'
    ) {
      if (curr.tagName === 'HTML') {
        // Fallback color if no element has a declared background-color attribute
        return undefined;
      }
      if (curr.parentElement) {
        curr = curr.parentElement;
      }
    }
    return window.getComputedStyle(curr).getPropertyValue('background-color');
  }
}

function _isOffsetHeightDifferent(a: React.RefObject<HTMLElement>, b: React.RefObject<HTMLDivElement>): boolean {
  return (a && b && a.current && b.current && a.current.offsetHeight !== b.current.offsetHeight) as boolean;
}
