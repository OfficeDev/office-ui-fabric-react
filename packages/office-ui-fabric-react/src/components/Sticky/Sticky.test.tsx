import * as React from 'react';
import { mount } from 'enzyme';
import { DetailsList, IDetailsHeaderProps } from 'office-ui-fabric-react/lib/DetailsList';
import { Sticky, StickyPositionType } from 'office-ui-fabric-react/lib/Sticky';
import { ScrollablePane } from 'office-ui-fabric-react/lib/ScrollablePane';
import { Breadcrumb } from 'office-ui-fabric-react/lib/Breadcrumb';
import { IRenderFunction } from 'office-ui-fabric-react/lib/Utilities';

// Populate mock data for testing
function mockData(count: number, isColumn: boolean = false, customDivider: boolean = false): any {
  const data = [];
  let _data = {};

  for (let i = 0; i < count; i++) {
    _data = {
      key: i,
      name: 'Item ' + i,
      value: i
    };
    if (isColumn) {
      _data = {
        ..._data,
        key: `column_key_${i}`,
        ariaLabel: `column_${i}`
      };
    }
    data.push(_data);
  }

  return data;
}

function onRenderDetailsHeader(props: IDetailsHeaderProps, defaultRender?: IRenderFunction<IDetailsHeaderProps>): JSX.Element {
  return (
    <Sticky stickyPosition={StickyPositionType.Header} isScrollSynced={true} stickyClassName={'details-header--sticky-test'}>
      {defaultRender!({
        ...props
      })}
    </Sticky>
  );
}

describe('Sticky', () => {
  it('sticky components are rendered in correct order when there is Breadcrumb', () => {
    const breadcrumbStyle = {
      root: [
        {
          margin: '0 0 0 0'
        }
      ]
    };
    const component = mount(
      <ScrollablePane className={'scrollable-pane-test'}>
        <Sticky stickyPosition={StickyPositionType.Header} stickyClassName={'breadcrumb--sticky-test'}>
          <Breadcrumb
            items={[
              { text: 'Files', key: 'Files' },
              { text: 'This is folder 1', key: 'f1' },
              { text: 'This is folder 2', key: 'f2' },
              { text: 'This is folder 3', key: 'f3' },
              { text: 'This is folder 4', key: 'f4' },
              { text: 'This is folder 5', key: 'f5', isCurrentItem: true }
            ]}
            ariaLabel={'breadcrumb-test'}
            styles={breadcrumbStyle}
          />
        </Sticky>
        <DetailsList
          items={mockData(5)}
          columns={mockData(5, true)}
          // tslint:disable-next-line:jsx-no-lambda
          onRenderRow={() => null}
          skipViewportMeasures={true}
          // tslint:disable-next-line:jsx-no-lambda
          onShouldVirtualize={() => false}
          onRenderDetailsHeader={onRenderDetailsHeader}
        />
      </ScrollablePane>
    );
    expect(component.html()).toMatchSnapshot();
  });
});
