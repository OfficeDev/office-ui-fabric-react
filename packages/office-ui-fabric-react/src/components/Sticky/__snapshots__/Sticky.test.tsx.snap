// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sticky sticky components are rendered in correct order when there is Breadcrumb 1`] = `

  <div
  class="ms-ScrollablePane
  scrollable-pane-test
  root-38"><div
  class="ms-ScrollablePane--contentContainer
  contentContainer-39"
  data-is-scrollable="true"><div><div
  style="height:
  0px;"></div></div><div
  class="ms-Viewport"
  style="min-width:
  1px;
  min-height:
  1px;"><div
  class="ms-DetailsList
  is-horizontalConstrained
  root-54"
  data-automationid="DetailsList"
  data-is-scrollable="false"><div
  role="grid"
  aria-rowcount="6"
  aria-colcount="6"
  aria-readonly="true"><div
  role="presentation"
  class="ms-DetailsList-headerWrapper
  headerWrapper-56"><div><div
  style="height:
  0px;"></div></div></div><div
  role="presentation"
  class="ms-DetailsList-contentWrapper
  contentWrapper-57"><div
  role="presentation"
  class="ms-FocusZone
  focusZone-55"
  data-focuszone-id="FocusZone9"><div
  class="ms-SelectionZone"
  role="presentation"><div
  role="presentation"
  class="ms-List"><div
  class="ms-List-surface"
  role="presentation"><div
  class="ms-List-page"
  role="presentation"><div
  role="presentation"
  class="ms-List-cell"
  data-list-index="0"
  data-automationid="ListCell"></div><div
  role="presentation"
  class="ms-List-cell"
  data-list-index="1"
  data-automationid="ListCell"></div><div
  role="presentation"
  class="ms-List-cell"
  data-list-index="2"
  data-automationid="ListCell"></div><div
  role="presentation"
  class="ms-List-cell"
  data-list-index="3"
  data-automationid="ListCell"></div><div
  role="presentation"
  class="ms-List-cell"
  data-list-index="4"
  data-automationid="ListCell"></div></div></div></div></div></div></div></div></div></div></div><div
  class="stickyAbove-40"
  style="height:
  0px;
  left:
  0px;
  right:
  0px;
  top:
  0px;"><div
  aria-hidden="false"
  style="pointer-events:
  auto;"><div
  style="visibility:
  hidden;
  height:
  0px;"></div><div
  class="breadcrumb--sticky-test"
  style="overflow:
  hidden;"><div><div
  style="position:
  relative;"><div
  style="position:
  fixed;
  visibility:
  hidden;"
  data-automation-id="visibleContent"><div
  class="ms-Breadcrumb
  root-43"
  role="navigation"
  aria-label="breadcrumb-test"><div
  role="presentation"
  class="ms-FocusZone"
  data-focuszone-id="FocusZone0"><ol
  class="ms-Breadcrumb-list
  list-44"><li
  class="ms-Breadcrumb-listItem
  listItem-45"><span
  class="ms-Breadcrumb-item
  item-50"><div
  class="ms-TooltipHost
  root-51">Files</div></span><i
  data-icon-name="ChevronRight"
  class="ms-Breadcrumb-chevron
  chevron-53"
  role="presentation"
  aria-hidden="true"></i></li><li
  class="ms-Breadcrumb-listItem
  listItem-45"><span
  class="ms-Breadcrumb-item
  item-50"><div
  class="ms-TooltipHost
  root-51">This
  is
  folder
  1</div></span><i
  data-icon-name="ChevronRight"
  class="ms-Breadcrumb-chevron
  chevron-53"
  role="presentation"
  aria-hidden="true"></i></li><li
  class="ms-Breadcrumb-listItem
  listItem-45"><span
  class="ms-Breadcrumb-item
  item-50"><div
  class="ms-TooltipHost
  root-51">This
  is
  folder
  2</div></span><i
  data-icon-name="ChevronRight"
  class="ms-Breadcrumb-chevron
  chevron-53"
  role="presentation"
  aria-hidden="true"></i></li><li
  class="ms-Breadcrumb-listItem
  listItem-45"><span
  class="ms-Breadcrumb-item
  item-50"><div
  class="ms-TooltipHost
  root-51">This
  is
  folder
  3</div></span><i
  data-icon-name="ChevronRight"
  class="ms-Breadcrumb-chevron
  chevron-53"
  role="presentation"
  aria-hidden="true"></i></li><li
  class="ms-Breadcrumb-listItem
  listItem-45"><span
  class="ms-Breadcrumb-item
  item-50"><div
  class="ms-TooltipHost
  root-51">This
  is
  folder
  4</div></span><i
  data-icon-name="ChevronRight"
  class="ms-Breadcrumb-chevron
  chevron-53"
  role="presentation"
  aria-hidden="true"></i></li><li
  class="ms-Breadcrumb-listItem
  listItem-45"><span
  class="ms-Breadcrumb-item
  item-50"><div
  class="ms-TooltipHost
  root-51">This
  is
  folder
  5</div></span></li></ol></div></div></div></div></div></div></div><div
  aria-hidden="false"
  style="pointer-events:
  auto;"><div
  style="visibility:
  hidden;
  height:
  0px;"></div><div
  class="details-header--sticky-test"
  style="overflow:
  hidden;"><div
  role="row"
  class="ms-FocusZone
  ms-DetailsHeader
  root-58"
  data-automationid="DetailsHeader"
  style="min-width:
  0;"
  data-focuszone-id="FocusZone8"><div
  class="ms-DetailsHeader-cell
  ms-DetailsHeader-cellIsCheck
  cellIsCheck-61"
  aria-labelledby="header7-check"
  aria-colindex="1"
  role="columnheader"><span
  class="checkTooltip-70"><div
  id="header7-check"
  data-is-focusable="true"
  role="checkbox"
  class="ms-DetailsRow-check
  ms-DetailsRow-check--isHeader
  checkHost-80"
  data-selection-toggle="true"
  data-automationid="DetailsRowCheck"
  tabindex="0"><div
  class="ms-Check
  root-77"><i
  data-icon-name="CircleRing"
  class="ms-Check-circle
  circle-83"
  role="presentation"
  aria-hidden="true"></i><i
  data-icon-name="StatusCircleCheckmark"
  class="ms-Check-check
  check-84"
  role="presentation"
  aria-hidden="true"></i></div></div></span></div><div
  role="columnheader"
  aria-sort="none"
  aria-colindex="2"
  class="ms-DetailsHeader-cell
  is-actionable
  root-87"
  data-is-draggable="false"
  draggable="false"
  style="width:
  120px;"
  data-automationid="ColumnsHeaderColumn"
  data-item-key="column_key_0"><span
  class="cellTooltip-95"><span
  id="header7-column_key_0"
  aria-labelledby="header7-column_key_0-name
  "
  class="ms-DetailsHeader-cellTitle
  cellTitle-93"
  data-is-focusable="true"
  aria-describedby="header7-column_key_0-tooltip"
  aria-haspopup="false"
  tabindex="-1"><span
  id="header7-column_key_0-name"
  class="ms-DetailsHeader-cellName
  cellName-94">Item
  0</span></span></span></div><label
  id="header7-column_key_0-tooltip"
  class="accessibleLabel-72">column_0</label><div
  role="columnheader"
  aria-sort="none"
  aria-colindex="3"
  class="ms-DetailsHeader-cell
  is-actionable
  root-87"
  data-is-draggable="false"
  draggable="false"
  style="width:
  120px;"
  data-automationid="ColumnsHeaderColumn"
  data-item-key="column_key_1"><span
  class="cellTooltip-95"><span
  id="header7-column_key_1"
  aria-labelledby="header7-column_key_1-name
  "
  class="ms-DetailsHeader-cellTitle
  cellTitle-93"
  data-is-focusable="true"
  aria-describedby="header7-column_key_1-tooltip"
  aria-haspopup="false"
  tabindex="-1"><span
  id="header7-column_key_1-name"
  class="ms-DetailsHeader-cellName
  cellName-94">Item
  1</span></span></span></div><label
  id="header7-column_key_1-tooltip"
  class="accessibleLabel-72">column_1</label><div
  role="columnheader"
  aria-sort="none"
  aria-colindex="4"
  class="ms-DetailsHeader-cell
  is-actionable
  root-87"
  data-is-draggable="false"
  draggable="false"
  style="width:
  120px;"
  data-automationid="ColumnsHeaderColumn"
  data-item-key="column_key_2"><span
  class="cellTooltip-95"><span
  id="header7-column_key_2"
  aria-labelledby="header7-column_key_2-name
  "
  class="ms-DetailsHeader-cellTitle
  cellTitle-93"
  data-is-focusable="true"
  aria-describedby="header7-column_key_2-tooltip"
  aria-haspopup="false"
  tabindex="-1"><span
  id="header7-column_key_2-name"
  class="ms-DetailsHeader-cellName
  cellName-94">Item
  2</span></span></span></div><label
  id="header7-column_key_2-tooltip"
  class="accessibleLabel-72">column_2</label><div
  role="columnheader"
  aria-sort="none"
  aria-colindex="5"
  class="ms-DetailsHeader-cell
  is-actionable
  root-87"
  data-is-draggable="false"
  draggable="false"
  style="width:
  120px;"
  data-automationid="ColumnsHeaderColumn"
  data-item-key="column_key_3"><span
  class="cellTooltip-95"><span
  id="header7-column_key_3"
  aria-labelledby="header7-column_key_3-name
  "
  class="ms-DetailsHeader-cellTitle
  cellTitle-93"
  data-is-focusable="true"
  aria-describedby="header7-column_key_3-tooltip"
  aria-haspopup="false"
  tabindex="-1"><span
  id="header7-column_key_3-name"
  class="ms-DetailsHeader-cellName
  cellName-94">Item
  3</span></span></span></div><label
  id="header7-column_key_3-tooltip"
  class="accessibleLabel-72">column_3</label><div
  role="columnheader"
  aria-sort="none"
  aria-colindex="6"
  class="ms-DetailsHeader-cell
  is-actionable
  root-87"
  data-is-draggable="false"
  draggable="false"
  style="width:
  120px;"
  data-automationid="ColumnsHeaderColumn"
  data-item-key="column_key_4"><span
  class="cellTooltip-95"><span
  id="header7-column_key_4"
  aria-labelledby="header7-column_key_4-name
  "
  class="ms-DetailsHeader-cellTitle
  cellTitle-93"
  data-is-focusable="true"
  aria-describedby="header7-column_key_4-tooltip"
  aria-haspopup="false"
  tabindex="-1"><span
  id="header7-column_key_4-name"
  class="ms-DetailsHeader-cellName
  cellName-94">Item
  4</span></span></span></div><label
  id="header7-column_key_4-tooltip"
  class="accessibleLabel-72">column_4</label></div></div></div></div><div
  class="stickyBelow-41"
  style="height:
  0px;
  left:
  0px;
  right:
  0px;
  bottom:
  0px;"><div
  class="stickyBelowItems-42"></div></div></div>
  {
    height: 32px;
  }
  .ms-Fabric--isFocusVisible &:focus {
    opacity: 1;
  }
  {
    -moz-osx-font-smoothing: grayscale;
    -webkit-font-smoothing: antialiased;
    align-items: center;
    background-color: transparent;
    background: none;
    border: none;
    box-sizing: border-box;
    cursor: default;
    display: flex;
    font-family: 'Segoe UI', 'Segoe UI Web (West European)', 'Segoe UI', -apple-system, BlinkMacSystemFont, 'Roboto', 'Helvetica Neue', sans-serif;
    font-size: 12px;
    font-weight: 400;
    height: 32px;
    justify-content: center;
    margin-bottom: 0px;
    margin-left: 0px;
    margin-right: 0px;
    margin-top: 0px;
    opacity: 0;
    outline: transparent;
    padding-bottom: 0px;
    padding-left: 0px;
    padding-right: 0px;
    padding-top: 0px;
    position: relative;
    vertical-align: top;
    width: 40px;
  }
  &::-moz-focus-inner {
    border: 0;
  }
  .ms-Fabric--isFocusVisible &:focus:after {
    border: 1px solid #ffffff;
    bottom: 1px;
    content: "";
    left: 1px;
    outline: 1px solid #666666;
    position: absolute;
    right: 1px;
    top: 1px;
    z-index: 1;
  }
  &:hover {
    opacity: 1;
  }
`;
